--- d/mm/vmscan.c~	2008-04-09 00:57:38.274635193 +0000
+++ d/mm/vmscan.c	2008-04-09 01:07:29.873975917 +0000
@@ -1004,7 +1004,7 @@ static void shrink_active_list(unsigned
 		 *
 		 * Max temporary value is vm_total_pages*100.
 		 */
-		imbalance *= (vm_swappiness + 1);
+		imbalance *= (vm_mapped + 1);
 		imbalance /= 100;
 
 		/*
@@ -1808,7 +1808,7 @@ void shrink_one_zone(struct zone *zone,
 		.gfp_mask = GFP_KERNEL,
 		.may_swap = 0,
 		.may_writepage = 1,
-		.swappiness = vm_swappiness,
+		.mapped = vm_mapped,
 	};
 
 	if (total_to_free <= 0)
