--- linux-2.6.18/drivers/usb/atm/cxacru.c.orig	2006-09-20 05:42:06.000000000 +0200
+++ linux-2.6.18/drivers/usb/atm/cxacru.c	2006-11-22 10:16:28.000000000 +0100
@@ -793,6 +793,9 @@
 	{ /* V = Conexant			P = ADSL modem				*/
 		USB_DEVICE(0x0572, 0xcb06),	.driver_info = (unsigned long) &cxacru_cb00
 	},
+	{ /* V = Conexant                       P = ADSL modem (ZTE ZXDSL 852)          */
+		USB_DEVICE(0x0572, 0xcb07),     .driver_info = (unsigned long) &cxacru_cb00
+	},
 	{ /* V = Olitec				P = ADSL modem version 2		*/
 		USB_DEVICE(0x08e3, 0x0100),	.driver_info = (unsigned long) &cxacru_cafe
 	},
