--- linux-2.6.17/include/linux/bio.h	2006-06-18 03:49:35.000000000 +0200
+++ linux-2.6.17.fcache/include/linux/bio.h	2006-06-21 15:58:45.000000000 +0200
@@ -124,6 +124,7 @@
 #define BIO_BOUNCED	5	/* bio is a bounce bio */
 #define BIO_USER_MAPPED 6	/* contains user pages */
 #define BIO_EOPNOTSUPP	7	/* not supported */
+#define BIO_NOMERGE	8	/* bio not mergeable */
 #define bio_flagged(bio, flag)	((bio)->bi_flags & (1 << (flag)))
 
 /*
