--- linux-2.6.18-ck1.orig/include/linux/sysctl.h	2006-09-21 20:03:59.000000000 +1000
+++ linux-2.6.18-ck1/include/linux/sysctl.h	2006-09-21 20:05:38.000000000 +1000
@@ -150,6 +150,9 @@ enum
 	KERN_IA64_UNALIGNED=72, /* int: ia64 unaligned userland trap enable */
 	KERN_COMPAT_LOG=73,	/* int: print compat layer  messages */
 	KERN_MAX_LOCK_DEPTH=74,
+	KERN_INTERACTIVE=75,	/* interactive tasks can have cpu bursts */
+	KERN_COMPUTE=76,	/* adjust timeslices for a compute server */
+	KERN_ISO_CPU=77,	/* percent cpu SCHED_ISO tasks run SCHED_RR */
 };
 
 
