--- linux-2.6.36/mm/swapfile.c.orig	2010-10-25 12:15:22.757143135 +0200
+++ linux-2.6.36/mm/swapfile.c	2010-10-25 12:16:10.084743007 +0200
@@ -499,6 +499,7 @@
 	spin_unlock(&swap_lock);
 	return (swp_entry_t) {0};
 }
+EXPORT_SYMBOL_GPL(get_swap_page_of_type);
 
 static struct swap_info_struct *swap_info_get(swp_entry_t entry)
 {
--- linux-2.6.36/scripts/mod/modpost.c.orig	2010-10-20 22:30:22.000000000 +0200
+++ linux-2.6.36/scripts/mod/modpost.c	2010-10-25 12:22:49.053556168 +0200
@@ -16,7 +16,8 @@
 #include <ctype.h>
 #include <string.h>
 #include "modpost.h"
-#include "../../include/generated/autoconf.h"
+// PLD architectures don't use CONFIG_SYMBOL_PREFIX
+//#include "../../include/generated/autoconf.h"
 #include "../../include/linux/license.h"
 
 /* Some toolchains use a `_' prefix for all user symbols. */

